### DISCUSSION TITLE
Start working on meow-tutor

### DISCUSSION THREAD

**Discussion by VitalyAnkh** [meow-edit/meow#86] [General]
Let's improve meow's documentation and make this great package newbie-friendly!

**Comment by VitalyAnkh** [comment]
My current plan: replicate a vim tutor for meow.
Draft here: https://github.com/meow-edit/meow/blob/cf53d48a13913f83ef247af43e864288d485ffb1/meow-tutor.el

**Reply by VitalyAnkh** [reply]
It's just a very draft. It's adapted from helix's tutor and I'm rewriting it. Just try revealing what a meow-tutor should do.

**Reply by DogLooksGood** [reply]
Will we go through all the features or just a subset. perhaps introducing too many things together, make it harder to get started.

**Reply by skmd0** [reply]
Examples of some of other tutors, maybe just for some additional ideas:

- [Kakoune Trampoline](https://github.com/mawww/kakoune/blob/master/contrib/TRAMPOLINE)
- [Boon tutorial](https://github.com/jyp/boon/blob/master/boon-tutorial.el)
- [Neovim Tutor](https://github.com/neovim/neovim/blob/c52ec8f9eb397da3440dcd821c41d5687b46ec16/runtime/tutor/en/vim-01-beginner.tutor)
- [Helix Tutor](https://github.com/helix-editor/helix/blob/master/runtime/tutor.txt)

**Reply by DogLooksGood** [reply]
The format of kakoune trampoline looks awesome. How about something like this?

**Reply by skmd0** [reply]
Yeah kakoune trampoline has awesome format, but it is missing exercises like in Neovim/Helix. The draft from @VitalyAnkh is really good start in my opinion. After the content gets written we can change the format anyway. I am thinking of starting to open a bunch of small PR for some fixes, or are you doing massive changes on it and I should wait a bit @VitalyAnkh ?

**Reply by DogLooksGood** [reply]
@skmd0 Agreed. I saw the draft is still not finished yet, let's wait a bit.

I've open new branch `meow-tutorial`.
@VitalyAnkh Please PR to the new branch.

**Reply by VitalyAnkh** [reply]
OK, I will do it tomorrow. :smile_cat:

**Reply by VitalyAnkh** [reply]
Hi all! I make it a elisp function, draft here: https://github.com/meow-edit/meow/blob/c7ce1276e00463e10cbb3d008498c635c4d04896/meow-tutor.el.

**Reply by Linerre** [reply]
If I'd like to contribute to tutorial stuff, which branch should I consider opening a PR to then? The new `meow-tutorial` branch or the `master`?

**Reply by DogLooksGood** [reply]
@Linerre we use `meow-tutorial`. you are in the meow-edit group, you can just push directly.

the same let you know @skmd0

**Reply by skmd0** [reply]
@DogLooksGood I am currently going over the meow-tutorial. Will submit changes after I am done.

**Reply by skmd0** [reply]
@VitalyAnkh Great job on writing the tutorial. 👍

I find it a great introduction for new people. I am currently fixing some grammatical mistakes and improving the brevity/phrasing of the content. I have made quite a few changes so I will open a PR after I am done.

**Reply by VitalyAnkh** [reply]
@skmd0 Thanks! Wating for your PR. :smile_cat:

**Comment by DogLooksGood** [comment]
Let me separate it into a text file and meow-tutor.el. 

simple text file is easy to maintain, and I can focus on meow-tutor.el.

**Reply by DogLooksGood** [reply]
pushed

**Reply by DogLooksGood** [reply]
since our recipe is not updated, so we can not extract it as a normal text file yet. 
just extract it into a const string.

**Reply by DogLooksGood** [reply]
The movement commands in meow can be divided into four groups

# basic movement, with explicit direction.
left/right/prev/next 
back-next/word-symbol

# regexp-search-ring related
mark-word/symbol (push)
visit (push)
search (use)

# common movements
line block find till reverse join

# thing-based movement
inner/bounds/beginning/end-of-thing

**Reply by DogLooksGood** [reply]
Putting too much content into the tutor is nod a good idea. so we can just pick a subset. 

Also, bmacro is avdanced usage, maybe it's not necessary to be in tutor.

**Reply by VitalyAnkh** [reply]
BMACRO is powerful! We should convince new users meow is powerful and worth using. We could just introduce a little bmacro.

**Comment by DogLooksGood** [comment]
Looks like we have our first version?

**Reply by skmd0** [reply]
Some sections in the very bottom are still to be rewritten if I understood correctly. 

Also some things that maybe should be added before first release:
- marking current word with (w) and extending selection
- unselect with g
- delete from cursor location until the end of line with s
- block/paragraphs movement/selection
- maybe some more concrete example on expending selection [ ] , .
- mention there is a cheatsheet and how to enable it
- keypad functionality mentioned

**Reply by DogLooksGood** [reply]
- a section for meow-mark-word/symbol, meow-visit and meow-search. mark word/symbol and visit can be used before search.
- unselect and reverse should be mentioned at beginning of MOTIONS AND SELECTIONS.
- a introduce for fallback commands? (meow-kill fallback to meow-C-k)
- meow-replace
- when exit from meow-change, content will be selected. so you can do: mark -> change -> type -> esc -> save -> search -> replace

**Reply by VitalyAnkh** [reply]
Note that we should keep the width of the text aligned.

**Reply by DogLooksGood** [reply]
@skmd0 will you add the additional contents?

**Reply by skmd0** [reply]
Sure I can try to write some of it.

**Reply by skmd0** [reply]
@DogLooksGood I saw the new PR. Do I push to meow-tutorial branch or do we merge first?

**Reply by DogLooksGood** [reply]
@skmd0 we merge when meow-tutor is complete. 
You can still commit after PR is opened.

**Reply by skmd0** [reply]
I have pushed all my changes into the tutorial branch. It took a very long time, but I am still not done. Will continue tomorrow.

I have noticed that even if the text content in lisp function is formatted nicely the output might not be, because `\\[meow-...]` is a single character in the output. We should probably aim for the final text in the tutor to have nicely aligned text.

The tutorial is becoming really good in my opinion!

**Reply by DogLooksGood** [reply]
I write a small command, hope it can help
```
(defvar meow-tutor-ovs nil)

(defun meow-tutor-toggle ()
  (interactive)
  (if meow-tutor-ovs
      (progn
        (mapc (lambda (o)
                (delete-overlay o))
              meow-tutor-ovs)
        (setq meow-tutor-ovs nil))
    (save-mark-and-excursion
      (goto-char (point-min))
      (while 
          (re-search-forward "\\\\\\\\\\[meow-.+?\\]" nil t)
        (let* ((m (match-data))
               (b (car m))
               (e (cadr m))
               (o (make-overlay b e)))
          (push o meow-tutor-ovs)
          (overlay-put o 'display "X")
          (overlay-put o 'face 'error))))))

(global-set-key (kbd "<f5>") 'meow-tutor-toggle)
```

press <kbd>F5</kbd> to toggle the display.

---

Great job! I'll commit for some minor changes and add `visit`, `search`.

**Reply by DogLooksGood** [reply]
I've finished sections for find/till, visit/search and keypad. 

I believe the tutorial is pretty good for now.

Please, add an ending section, and of course, something like `written by ...`

@VitalyAnkh @skmd0

**Reply by DogLooksGood** [reply]
Worth taking a look, there's another [great progress](https://raw.githubusercontent.com/Linerre/meow/scratch-post/SCRATCH_POST.org) by @Linerre 

This could be our next tutorial I think.

**Reply by skmd0** [reply]
Wow that is really cool format!

**Reply by DogLooksGood** [reply]
merged into master, now `M-x meow-tutor` is available.

**Comment by Fedreg** [comment]
Just thought I'd chime in as a new meow user...

I pulled in `meow-tutor` and it was REALLY helpful; great job! on it!!

the additional sections that @skmd0 mentioned would definitely be helpful to newcomers (but it's already really good even without...)

thx!!
